libs/framework/src/service_tracker.c

```diff
@@ -32,6 +32,8 @@
 #include "celix_log.h"
 #include "bundle_context_private.h"
 #include "celix_array_list.h"
+#include "../../utils/src/version_range_private.h"
```

Comment:
Do not include with relative paths and do not include "private" headers (headers in the src dirs).

I would prefer that a function for transforming a version ranges to ldap filter is added to the utils library:

typedef struct versionRange version_range_t;

/**
 * Returns the LDAP filter for a version range. Caller is owner of the returned string.
 * If version range is not valid, NULL will be returned.
 */
char* versionRange_createLDAPFilter(version_range_t *range)

/**
 * construct a LDAP filter for the provided version range. 
* The string will be created in the provided buffer, if the buffer is big enough.
*  Returns true if successful. 
 */
bool versionRange_constructLDAPFilter(version_range_t *range, char* buffer, size_t bufferLength);

Reply:
While I agree that including private headers is nasty, the source code in service_tracker.c needs the full definition of the structs in those headers, otherwise it simply does not compile. Would you be ok with removing the private header files and moving the definitions of the structs into version_range.h?